eidas_node/saml.py

```diff
@@ -0,0 +1,226 @@
+from lxml import etree
+from lxml.etree import Element, ElementTree, QName, SubElement
+
+from eidas_node.constants import StatusCode, SubStatusCode
+from eidas_node.errors import ValidationError
+from eidas_node.models import LevelOfAssurance, LightRequest, LightResponse, NameIdFormat, Status
+from eidas_node.utils import get_element_path
+
+NAMESPACES = OrderedDict([
+    ('saml2', 'urn:oasis:names:tc:SAML:2.0:assertion'),
+    ('saml2p', 'urn:oasis:names:tc:SAML:2.0:protocol'),
+    ('eidas', 'http://eidas.europa.eu/saml-extensions'),
+    ('xmlenc', 'http://www.w3.org/2001/04/xmlenc#'),
+])  # type: Dict[str, str]
+"""XML namespaces in SAML requests."""
+
+KNOWN_TAGS = {
+    'saml2': {'Issuer', 'AuthnContextClassRef', 'EncryptedAssertion', 'Assertion', 'Subject', 'NameID',
+              'AuthnStatement', 'AttributeStatement', 'Attribute', 'AttributeValue', 'SubjectLocality',
+              'AuthnContext', 'AuthnContextClassRef'},
+    'saml2p': {'AuthnRequest', 'Extensions', 'NameIDPolicy', 'RequestedAuthnContext',
+               'Response', 'Status', 'StatusCode', 'StatusMessage'},
+    'eidas': {'SPType', 'SPCountry', 'RequestedAttributes', 'RequestedAttribute', 'AttributeValue'},
+    'xmlenc': {'EncryptedData'}
+}  # type: Dict[str, Set[str]]
+"""Recognized XML tags in SAML requests."""
+
+Q_NAMES = {
+    '{}:{}'.format(ns, tag): QName(NAMESPACES[ns], tag) for ns, tags in KNOWN_TAGS.items() for tag in tags
+}  # type: Dict[str, QName]
+"""Qualified names of recognized XML tags in SAML requests."""
+
+SAMLRequestType = TypeVar('SAMLRequestType', bound='SAMLRequest')
+SAMLResponseType = TypeVar('SAMLResponseType', bound='SAMLResponse')
+
+
+class SAMLRequest:
+    """SAML Request and its conversion from/to LightRequest."""
+
+    document = None  # type: ElementTree
+    """SAML document as an element tree."""
+    relay_state = None  # type: Optional[str]
+    """Relay state associated with the request."""
+
+    def __init__(self, document: ElementTree, relay_state: Optional[str] = None):
+        self.document = document
+        self.relay_state = relay_state
+
+    @classmethod
+    def convert_light_request(cls: Type[SAMLRequestType], light_request: LightRequest,
+                              destination: str, issued: datetime) -> SAMLRequestType:
+        """
+        Convert Light Request to SAML Request.
+
+        :param light_request: The light request to convert.
+        :param destination: A URI reference indicating the address to which this request has been sent.
+        :param issued: The UTC time instant of issue of the request.
+        :return: A SAML Request.
+        """
+        light_request.validate()
+        root_attributes = OrderedDict([
+            ('Consent', 'urn:oasis:names:tc:SAML:2.0:consent:unspecified'),  # optional, default 'unspecified'
+            ('Destination', destination),
+            ('ID', light_request.id),
+            ('IssueInstant', issued.isoformat()[:-3] + 'Z'),  # UTC
```

Comment:
There is a function for this conversion already, isn't it?

Reply:
Do you mean the function for eIDAS date-time format conversion? The SAML format is different...

Reply:
OK, looked familiar...